### DISCUSSION TITLE
CustomDrawers or custom shaders for WebGLDrawer

### DISCUSSION THREAD

**Discussion by Tobio89** [openseadragon/openseadragon#2633] [Ideas]
Hi OpenSeadragon,

Recently I was working on a custom drawer implementation for my company. I needed to take control of the shader in WebGLDrawer so I could process colours in a specific way.

_I was wondering if there is any plan to allow developers to pass their own shaders into WebGLDrawer._

For my implementation, I needed to copy WebGLDrawer and then modify the shader. Since OSD only allows you to have one drawer in the viewer, I set a control uniform inside the shader to either use or ignore the additional shader code I wrote.

I then passed in custom data into the tiled image source.

It would be great if an official pipeline for that kind of operation could be set up!

**Comment by Tobio89** [comment]
Ah, Ian suggested I invite @pearcetm and @Aiosa to this discussion!

**Comment by Aiosa** [comment]
I have a student implementing webgl renderer that allows running custom shaders in a modular way, support both WebGL 1 and 2 and so on. He is submitting his bc thesis this winter, and I will be left with half-ready rendering solution, which needs to be optimized & cleaned. We could use this.

**Comment by pearcetm** [comment]
@Tobio89 Yeah, this is something I want to do as well. We may be able to start with the prototype that @Aiosa mentions (I haven't seen that yet, so we'll have to see how beneficial it would be vs other possible approaches).

The way I was conceptualizing it was that custom shader code could be passed into the `WebGLDrawer` and compiled into a shader program. It could then be attached to either the entire `drawer` for use in all `TiledImage`s, or could apply to just a single `TiledImage` (so that the default implementation could be used for others, or different custom shader programs per `TiledImage`). This may incur too much overhead of switching programs though, I'm not sure.

But yes, we should work on adding this functionality!

**Reply by Aiosa** [reply]
The submission deadline is next wednesday, so after that we can look into it.

**Reply by Tobio89** [reply]
I'd be quite excited for that! But it would also be great to be able to control uniforms etc too. 
DeckGL has a feature where you can specify uniforms in an object and it handles the actual instantiation of them behind the scenes. I wonder if something like that could work?

**Reply by Aiosa** [reply]
Stuff like this is simple. Problems lie in optimizing (it does not use instanced rendering RN) and blending. The hardest of all is postprocessing with Context2D, which would be good to avoid, but it's very hard to do stuff like text rendering on GPU, and people are used to access the viewport data using it's API too.

**Comment by Aiosa** [comment]
So I have the code, I started working on the integration logics (it does not yet support the latest cache API), but I arrived at at the fact that my latest API (https://github.com/openseadragon/openseadragon/pull/2643) is not yet merged, and I don't want to program something I know has a change proposal and throw everything away. So I will wait for the PR before I submit the code..

**Comment by Aiosa** [comment]
Okay, I am done. The implementation base is here: https://github.com/RationAI/openseadragon/tree/feat/modular-renderer

### How it works

Just run the demo comparison page and select 'modular renderer' in the options below. New options will be available to choose from (e.g. the shader type to apply).

Each tiled image configures a shader (and its properties like what filters to apply, what channels to sample) via a JSON configuration map. Shaders have API that ensures easy itnegration & naive version-less support.

### What needs improvements
 - **Optimization.** currently, there is no optimalization like instanced rendering used, which looks like a must.
 - **Custom Blending.** optimalization will mean that with two-pass rendering, the layer blending happens on the fragment shader level. The implementation is there, it's just not being used. Also layer clipping should be supported.
 - **Found a bug.** Currently, OSD does not pass drawer options to the navigator, which results in different behavior.
 - **True glsl version-less support.** Current WebGL context only creates a proxy functions (e.g. texture access), but does not ensure full compatibility. E.g. if you use locally defined arrays, WebGL 1.0 will not render since it uses webgl2.0 only compatible syntex. So when you program a shader, you still have to be careful about using only code that works in both versions. Which is not THAT limiting (in the context of our usecase), but still. -- That means the current demo of Sobel shader works only on Webgl 2.0.
 - **More modularity.** Shaders could be actually individual modules that can be connected together to form a rendering pipeline. This way, instead of current rather 'dirty' way of defining preprocessing filters (like gamma), we could create our shaders from such modules. Think of a visual programming-like design.
 - **More simplicity.** I already removed some advanced features, we could even more remove stuff and leave the basic API very simple, with moving all the advanced stuff to a plugin.

![image](https://github.com/user-attachments/assets/1eeeabcf-c4d6-492d-b986-2bfeb4787e72)

**Reply by iangilman** [reply]
@Aiosa This looks fantastic! I'm curious to hear what @Tobio89 and @pearcetm think 😊 0

**Comment by Aiosa** [comment]
Just FYI I was working lately quite extensively on this and I have working, optimized renderer that supports all except composite + debug info. There is still a LOT of work (it only uses webgl2, will be probably broken on apple due to their implementation bug), but it is already quite usable.

I am planning to also add the webgl1 counterpart that will use the context2d processing  for stuff that is too hard to do on gpu, or working with tainted data; and will be penalized therefore in speed.

**Reply by iangilman** [reply]
Awesome. Thank you for the update. This is exciting stuff!

**Comment by Aiosa** [comment]
A small taste (composite added):

Rainbow rendered as-is.
Leaves are rendered with sobel shader & darken blending. With clipping.
Letter 'B' is rendered as-is.
Duomo is cropped, rendered using 'mask' blending (erases all data on 'background' color where it does not render itself) + applied 'clipping' mode -> applies only on previous layer (background = letter B). E.g. we use Duomo layer to crop its geometry on the letter B itself, as if we selected crop on the B layer instead.

![image](https://github.com/user-attachments/assets/31d00736-7206-4341-9a91-992bf297f6f1)

There are still conrner case bugs, and not all scenarios are supported (debugging info, webgl1). Everything runs on GPU, no canvas2d API used.

Edit: forgot to add duomo settings:
![image](https://github.com/user-attachments/assets/ed93e616-b2c6-439f-b880-d5c9a8e9bff6)

**Reply by iangilman** [reply]
Awesome. Very powerful!

**Comment by Aiosa** [comment]
Update: apart from some minor compability issues (like debugging support), the renderer is already working quite well, except for performance. It's not bad but also not that great. 

Inspecting the cause, the animation frame is by >60% occupied by update levels call on tiled images. The new renderer is also not too bad, but takes ~30/40 ms to render, which adds to already slow frame handler, and if GC decides to show up on the party ...

The problem I struggle with is that I need to render also clipping textures alongside texture sewing itself, which forces me to iterate over tiled images and render them separately. Simply, the sewing is too slow to my linking. The renderer features also make it quite cumbersome to support single pass rendering, I mean if you don't need any of these features just use one of the existing renderers.

I have successfully integrated the renderer within my pathology viewer, RN I focus on some minor jumpy artifacts and the optimization... I am also thinking about looking about the update routine logics, we could cache some computations (like tiled image position and bounds which might be computed several times each frame), and also consider using threads via Workers. That might be harder given that code on a worker needs to have all of the logics loaded separately. For example we could try to separate image loader, but it invokes tile source API to download data - it brings flexibility when you need custom handling, but also makes such optimization hard...

**Reply by iangilman** [reply]
Thank you for the update. You're really tangling with a lot of challenging problems! ❤️

**Comment by Aiosa** [comment]
So I tried to optimize the heck out of WebGL 2 API and either I am missing some bugs, having some misunderstandings, or just arriving at the limits. Anyway, it is sort of optimized, though the performance could be better IMO. RN I don't see a simple way of improving it further, though I want to add a few more features (like support for 3d texture asset loading which would map later to for example icc profile mapping usecase).

I was thinking about trying web workers, trying to push most of OSD logic to a worker thread and running only drawers on the UI thread. Though it is not that simple, there are still features that would have to be factored out of the core codebase to draw an exact line between what is running on the backend and what is touching the UI (overlays, reference strip, buttons...). Might be a nice idea to try this in the (near?) future.

So should I open PR? Do we want to have this renderer in OSD or as a standalone plugin?

**Reply by iangilman** [reply]
Bravo! 

How would you describe this renderer, in comparison to the existing WebGL and canvas drawers? How possible is it to make it a plugin vs. built in?

**Reply by Aiosa** [reply]
It's closer to WebGL renderer from which I've started. It does not use context2d api to deliver the functionality:
 - clipping polygons are working only when they are convex
    - this is just to avoid adding external libraries, earcut.js or similar is otherwise necessary
 - no debugging mode so far
 - no support for tained data rendering
    - there is no CPU processing involved as the point is to render shaders, using e.g. context2d does not preserve the funtionality
 - it has manual blending implemented on fragment shader 
    - proven fix to some issues WebGL renderer has with PNG data
 - it has two modes, you can in one mode override the whole rendering logics by custom configuration: you can add any shader and read from any tiled image, so you can render in any order
 - it is implemented in WebGL2 (impossible to do the previous feature otherwise, but WebGL 1 could work too except you could not read from ANY layer and the performance is potentially worse)
 - it has advanced rendering capabilities (like clipping mask) and keeps mask of written pixels, so that e.g. inversion does not render to pixels outside the tiled image bounds
 - and more less important aspects
 
Missing support for rendering tainted data & drawing debug info could be implemented by WebGL1 (it has support for seamless version switching) at the cost of sacrificing some performance. I would say that there is almost no difference in plugin VS core renderer setup, except with plugin people have to make sure the load order of libs is `osd` -> `plugin` -> `custom code` which is obvious..

**Reply by iangilman** [reply]
Cool. Sounds like it's a bit more specialized than the current WebGL renderer. Sounds cool!

How about we start with it as a plugin and see how it goes? Having a drawer plugin is probably a good step for us, now that it's possible. Over time we may bring aspect into the core, or bring the whole thing in, depending on how things evolve.

@pearcetm Any thoughts?

**Reply by pearcetm** [reply]
I agree that this is a nice extension, and will work well for some things but not others. Makes sense to me for it to be a plugin. I believe you can pass the constructor of a class that inherits from `DrawerBase` directly as the option for `drawer` when creating a viewer.

**Reply by Aiosa** [reply]
No need for that. If you create the drawer class in OSD namespace you can pass the type string as well.

**Comment by rgov** [comment]
@Aiosa I tested your commit 49ff37e2962d1707af6ecc4d20de62727148647e on Safari and it works pretty well. 9205f16fff8224e6ed71fbb1e6843be4368c835a seems to add some debugging code and I couldn't get the Sobel filter to work.

My intended application of this is for a technique like bump mapping, with tiles that include a surface normal channel. I'm not experienced with GLSL but it looks like I could maybe have two TiledImages, one for my RGB channels and one for my normals channel, and maybe a custom blend function?

**Reply by Aiosa** [reply]
Thanks for info. I will need to polish this a bit and release it, I was just busy two weeks on a row now (summer camp & conference) so I will surely be more active next week :)

**Reply by Aiosa** [reply]
@rgov I am back, but still a bit ill so it might take me some time to finish the drawer. Btw instead of custom 'blend' function it should IMHO just create a new shaderlayer that takes for input two image sources... Totally valid.

**Comment by Aiosa** [comment]
First implementation has been released here: https://github.com/openseadragon/flex-render 
It might take some time to polish it yet and breaking changes are possible :)

Note though, that you have to:
 - use latest OSD v6 (current master)
 - <strike>have bugfixes from this PR: https://github.com/openseadragon/openseadragon/pull/2762</strike> (merged)

**Reply by iangilman** [reply]
Awesome! You should share this on the Discord as well 😊

**Reply by Aiosa** [reply]
I will as long as we progress with the PR, so people can use current master with the renderer.

**Reply by iangilman** [reply]
Sounds good. Apologies for the slow down!
